crates/ruff_python_semantic/src/statements.rs

```diff
@@ -0,0 +1,100 @@
+use std::ops::{Index, IndexMut};
+
+use ruff_index::{newtype_index, IndexVec};
+use ruff_python_ast::Stmt;
+use rustc_hash::FxHashMap;
+
+use ruff_python_ast::types::RefEquality;
+
+/// Id uniquely identifying a statement AST node.
+///
+/// Using a `u32` is sufficient because Ruff only supports parsing documents with a size of max
+/// `u32::max` and it is impossible to have more nodes than characters in the file. We use a
+/// `NonZeroU32` to take advantage of memory layout optimizations.
+#[newtype_index]
+#[derive(Ord, PartialOrd)]
+pub struct StatementId;
+
+/// A [`Stmt`] AST node, along with a pointer to its parent statement (if any).
+#[derive(Debug)]
+struct Statement<'a> {
+    /// A pointer to the AST node.
+    node: &'a Stmt,
+    /// The ID of the parent of this node, if any.
+    parent: Option<StatementId>,
+    /// The depth of this node in the tree.
+    depth: u32,
```

Comment:
Is this a frequent operation or could we compute the `depth` by counting the ancestor chain`?

Reply:
I will revisit in a follow-up PR, since this already exists on `Nodes` (now removed).